ui/components/app/assets/token-list/token-list.tsx

```diff
@@ -8,73 +9,202 @@ import {
+  useEffect(() => {
+    const allNetworkFilters = Object.fromEntries(
+      Object.keys(allNetworks).map((chainId) => [chainId, true]),
+    );
+
+    if (Object.keys(tokenNetworkFilter).length > 1) {
+      dispatch(setTokenNetworkFilter(allNetworkFilters));
+    }
+  }, [Object.keys(allNetworks).length]);
+
+  const consolidatedBalances = () => {
+    const tokensWithBalance: TokenWithFiatAmount[] = [];
+
+    Object.entries(selectedAccountTokensChains).forEach(
+      ([stringChainKey, tokens]) => {
+        const chainId = stringChainKey as Hex;
+        tokens.forEach((token: Token) => {
+          const { isNative, address, decimals } = token;
+          const balance =
+            calculateTokenBalance({
+              isNative,
+              chainId,
+              address,
+              decimals,
+              nativeBalances,
+              selectedAccountTokenBalancesAcrossChains,
+            }) || '0';
+
+          const tokenFiatAmount = calculateTokenFiatAmount({
+            token,
+            chainId,
+            balance,
+            marketData,
+            currencyRates,
+          });
+
+          // Append processed token with balance and fiat amount
+          tokensWithBalance.push({
+            ...token,
+            balance,
+            tokenFiatAmount,
+            chainId,
+            string: String(balance),
+          });
+        });
+      },
+    );
+
+    return tokensWithBalance;
   };
 
-  const sortedTokens = useMemo(() => {
-    // TODO filter assets by networkTokenFilter before sorting
-    return sortAssets(
-      [nativeTokenWithBalance, ...tokensWithBalances],
-      tokenSortConfig,
+  const sortedFilteredTokens = useMemo(() => {
+    const consolidatedTokensWithBalances = consolidatedBalances();
+    const filteredAssets = filterAssets(consolidatedTokensWithBalances, [
+      {
+        key: 'chainId',
+        opts: tokenNetworkFilter,
+        filterCallback: 'inclusive',
+      },
+    ]);
+
+    const { nativeTokens, nonNativeTokens } = filteredAssets.reduce<{
+      nativeTokens: TokenWithFiatAmount[];
+      nonNativeTokens: TokenWithFiatAmount[];
+    }>(
+      (acc, token) => {
+        if (token.isNative) {
+          acc.nativeTokens.push(token);
+        } else {
+          acc.nonNativeTokens.push(token);
+        }
+        return acc;
+      },
+      { nativeTokens: [], nonNativeTokens: [] },
     );
+    const assets = [...nativeTokens, ...nonNativeTokens];
+    return sortAssets(assets, tokenSortConfig);
   }, [
-    tokensWithBalances,
     tokenSortConfig,
     tokenNetworkFilter,
     conversionRate,
     contractExchangeRates,
+    currentNetwork,
+    selectedAccount,
+    selectedAccountTokensChains,
+    newTokensImported,
   ]);
 
   useEffect(() => {
-    if (!loading) {
+    if (sortedFilteredTokens) {
       endTrace({ name: TraceName.AccountOverviewAssetListTab });
     }
-  }, [loading]);
+  }, [sortedFilteredTokens]);
 
+  const loading = false;
```

Comment:
This isn't helpful considering the next line.

Reply:
Thanks, removing now

Reply:
https://github.com/MetaMask/metamask-extension/pull/28386/commits/a9b16e2b316645e3b1530dbf1fece5b1ce45efa3